runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/test/aws/smithy/kotlin/runtime/http/engine/okhttp/ByteChannelRequestBodyTest.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.http.engine.okhttp
+
+import aws.smithy.kotlin.runtime.hashing.sha256
+import aws.smithy.kotlin.runtime.http.HttpBody
+import aws.smithy.kotlin.runtime.io.SdkByteChannel
+import aws.smithy.kotlin.runtime.io.SdkByteReadChannel
+import aws.smithy.kotlin.runtime.util.encodeToHex
+import kotlinx.coroutines.*
+import kotlinx.coroutines.test.runTest
+import okio.Buffer
+import org.junit.jupiter.api.Test
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.test.assertEquals
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+@OptIn(ExperimentalCoroutinesApi::class)
+class ByteChannelRequestBodyTest {
+    @Test
+    fun testWriteTo() = runTest {
+        val content = ByteArray(1024 * 12 + 13) { it.toByte() }
+        val expectedSha256 = content.sha256().encodeToHex()
+        val chan = SdkByteReadChannel(content)
+        val body = object : HttpBody.Streaming() {
+            override val contentLength: Long = content.size.toLong()
+            override fun readFrom(): SdkByteReadChannel = chan
+        }
+
+        val callContext = coroutineContext + Job()
+        val actual = ByteChannelRequestBody(body, callContext)
+
+        assertEquals(body.contentLength, actual.contentLength())
+        assertFalse(actual.isDuplex())
+        assertTrue(actual.isOneShot())
+
+        val buffer = Buffer()
+        actual.writeTo(buffer)
+        val actualSha256 = buffer.sha256().hex()
+        assertEquals(expectedSha256, actualSha256)
```

Comment:
**Question**: Why compare the hashes? Why not just compare the bytes themselves?

Reply:
That would work but this matches the rest of our "integrity" style checks scattered in the runtime tests. It also encapsulates two checks in one, are they the same length and are the contents the same?